### PR TITLE
Fix S3358 FP: Nested ternary operator is in a lambda

### LINKED ISSUE #5773: Fix S3358 FP: Nested ternary operator is in a lambda
### Description

see https://community.sonarsource.com/t/csharpsquid-s3358-considers-lambdas-for-nesting/66955

Don't report here:
```cs
    return withTimeAccessor != null
        ? item => withTimeAccessor(item) ? ReportCellStyle.DateTime : ReportCellStyle.Date
        : item => withTime ? ReportCellStyle.DateTime : ReportCellStyle.Date;
```

### Justification

1. Different context/semantics

The lambda is a new context. A ternary is problematic when the true and/or false expressions are ternary expressions. This is not the case here. Consider e.g.

```cs
var gamble = playersAge >= 18
  ? diceRolls.Select(x => x % 2 == 0 ? "You win": "You loose")
  : Enumerable.Empty<string>();
```
 The two ternaries are running in completely different contexts.

2. Re-factoring

Nested ternaries can be refactored, e.g. like so

```cs
// Before
var greeting = isMale
  ? "Mr."
  : isMarried
    ? "Mrs."
    : "Miss";

// After
var greeting = (isMale, isMarried) switch
  {
    (true, _) => "Mr",
    (_, true) => "Mrs.",
    (_, false) => "Miss",
  };
```

I don't see a straightforward way to get rid of the nesting in the case of lambdas.

### PR DESCRIPTION
**Pull request by martin-strecker-sonarsource**
Fixes #5773

Todo

* [x] VB Support?

### CODE REVIEW

**Comment by Corniel** [comment]
As mentioned at the [Community Platform](https://community.sonarsource.com/t/csharpsquid-s3358-considers-lambdas-for-nesting/66955), I think this is not a false positive; it's about readability, and that does not change by having lamda's or not.

**Comment by martin-strecker-sonarsource** [comment]
@Corniel Readability certainly is one aspect of the rule.  I think this is an FP for these reasons:

1. Different context/semantics

The lambda is a new context. A ternary is problematic when the true and/or false expressions are ternary expressions. This is not the case here. Consider e.g.

```cs
var gamble = playersAge >= 18
  ? diceRolls.Select(x => x % 2 == 0 ? "You win": "You loose")
  : Enumerable.Empty<string>();
```
 The two ternaries are running in completely different contexts.

2. Re-factoring

Nested ternaries can be refactored, e.g. like so

```cs
// Before
var greeting = isMale
  ? "Mr."
  : isMarried
    ? "Mrs."
    : "Miss";

// After
var greeting = (isMale, isMarried) switch
  {
    (true, _) => "Mr",
    (_, true) => "Mrs.",
    (_, false) => "Miss",
  };
```

I don't see a straightforward way to get rid of the nesting in the case of lambdas.

**Comment by martin-strecker-sonarsource** [comment]
The ITs found the following FPs
```cs
// 1.
                var configWithDefaultFallback = bootstrap.HasValue
                    ? bootstrap.Value.Config.Select(c => c == _defaultConfig ? c : c.WithFallback(_defaultConfig))
                    : _defaultConfig;
// 2.
            var messages = (SupportsAtomicPersistAllOfSeveralEvents
                ? Enumerable.Range(from, to - 1)
                    .Select(
                        i =>
                            i == to - 1
                                ? new AtomicWrite(
                                    new[] {persistent(i), persistent(i + 1)}.ToImmutableList<IPersistentRepresentation>())
                                : new AtomicWrite(persistent(i)))
                : Enumerable.Range(from, to).Select(i => new AtomicWrite(persistent(i))))
                .ToArray();

// 3.
            return empty != null
                ? new object[0]
                : ctors.First().GetParameters().Select(p => p.ParameterType.IsValueType ? Activator.CreateInstance(p.ParameterType) : null).ToArray();

```

File: analyzers/its/expected/akka.net/Akka.MultiNodeTestRunner--net471-S3358.json [hunk 1]
```diff
@@ -1,17 +0,0 @@
-{
-"issues":  [
-{
-"id":  "S3358",
-"message":  "Extract this nested ternary operation into an independent statement.",
-"location":  {
-"uri":  "sources\akka.net\src\core\Akka.MultiNodeTestRunner\Discovery.cs",
```
Comment (pavel-mikula-sonarsource) [inline analyzers/its/expected/akka.net/Akka.MultiNodeTestRunner--net471-S3358.json:7]: Nice

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/DoNotNestTernaryOperators.cs [hunk 2]
```diff
@@ -24,6 +24,14 @@ public bool FooImpl(bool isMale, bool isMarried)
                 ? "Mrs. "
                 : "Miss ";
 
+            var lambda1 = true // Compliant. Nested in Lambda is valid
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/DoNotNestTernaryOperators.cs:27]: Let's add noncompliant 
```
var lambda3 = new Func<string, string>(s => isMale ? "Mr. " : isMarried ? "Mrs. " : "Miss ")
```

**Review by pavel-mikula-sonarsource — APPROVED** [review]
LGTM

**Comment by martin-strecker-sonarsource** [comment]
Validation errors:
* https://peach.sonarsource.com/project/issues?id=aspnetcore&issues=AXoJkZ-rXaALAbC4waMZ&open=AXoJkZ-rXaALAbC4waMZ

**Comment by martin-strecker-sonarsource** [comment]
@andrei-epure-sonarsource I checked this on peach and found one nonreproducible issue (see above). Can we have a look together?

**Pull request merged**